File: C++/2-Medium/00264._丑数_II/00264._丑数_II.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int nthUglyNumber(int n) {
    vector<int> dp(n, 0);
    dp[0] = 1; // 第一个丑数为 1
    int p2 = 0, p3 = 0, p5 = 0; // 初始化三个指针
    
    for (int i = 1; i < n; ++i) {
        // 计算下一个丑数
        int next2 = dp[p2] * 2, next3 = dp[p3] * 3, next5 = dp[p5] * 5;
        dp[i] = min(next2, min(next3, next5));
        
        // 更新指针
        if (dp[i] == next2) ++p2;
        if (dp[i] == next3) ++p3;
        if (dp[i] == next5) ++p5;
    }
    return dp[n - 1];
}

int main() {
    int n;
    cout << "请输入 n 的值: ";
    cin >> n;
    cout << "第 " << n << " 个丑数是: " << nthUglyNumber(n) << endl;
    return 0;
}

File: Python/2-Medium/00264._丑数_II/00264._丑数_II.py
from typing import List

def nthUglyNumber(n: int) -> int:
    """
    寻找第 n 个丑数。
    
    参数:
        n (int): 要找的丑数的序号。
        
    返回:
        int: 第 n 个丑数。
        
    示例:
        >>> nthUglyNumber(10)
        12
    """
    ugly = [1] * n  # 存储丑数的数组
    i2 = i3 = i5 = 0  # 初始化三个指针
    
    for i in range(1, n):
        # 分别计算乘以 2、3、5 的结果，并取最小值
        next_ugly = min(ugly[i2] * 2, ugly[i3] * 3, ugly[i5] * 5)
        ugly[i] = next_ugly
        
        # 如果当前丑数是由某个指针对应的值乘以对应的因子得到的，则该指针前进一步
        if next_ugly == ugly[i2] * 2: i2 += 1
        if next_ugly == ugly[i3] * 3: i3 += 1
        if next_ugly == ugly[i5] * 5: i5 += 1
    
    return ugly[-1]  # 返回第 n 个丑数

# 运行代码，以 n = 10 为例
print(nthUglyNumber(10))